### PR TITLE
get compile_colmmands.json support started

### LINKED ISSUE #101: Support meson projects
Provide auto completion of files as specified in the meson file + dependencies

cc @Vild

### LINKED ISSUE #110: Support compile_commands.json
Some build systems such as CMake and Meson issue a `compile_commands.json` file intended to be read by linters.
This could be used as alternative to Dub to provide the import paths and other options.

The `compile_commands.json` provide one entry per source file with the build options:
```json
[
  {
    "directory": "C:\\sdk\\gfx-d\\builddir",
    "command": "\"ldc2\" \"-I=vulkan\\libgfxd_vulkan.a.p\" \"-I=vulkan\" \"-I=..\\vulkan\" \"-I=..\\vulkan\\source\" \"-I=..\\graal\\source\" \"-I=..\\core\\source\" \"-I=..\\math\\source\" \"-enable-color\" \"-wi\" \"-m64\" \"-g\" \"-d-debug\" \"-relocation-model=pic\" -of=vulkan/libgfxd_vulkan.a.p/source_gfx_vulkan_device.d.obj \"-c\" ../vulkan/source/gfx/vulkan/device.d",
    "file": "../vulkan/source/gfx/vulkan/device.d",
    "output": "vulkan/libgfxd_vulkan.a.p/source_gfx_vulkan_device.d.obj"
  },
  ...
]
```

It should be quite easy to parse the command field and extract the relevant import dirs, versions, debug flags etc.

### PR DESCRIPTION
**Pull request by rtbo**
Not anywhere near something functional, but I'm posting to start to get guidance.
Currently I've made a component that reads the `compile_commands.json` file and extract import paths, versions etc.

Not sure what else is required to have functional linting, if we don't have Dub at all.
Note that I don't intend to drive the build from the extension, only to provide linting.

Ultimately, I don't necessarily want to extract import paths, versions etc, but to actually run the command associated with the open file each time it is modified and report back the errors from the compiler.

fixes #101
fixes #110

### CODE REVIEW

File: source/workspaced/com/ccdb.d [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+		ref HashSet!string stringImports,
+		ref HashSet!string fileImports,
+		ref HashSet!string versions,
+		ref HashSet!string debugVersions)
+	{
+		import std.algorithm : startsWith;
+
+		enum importMark = "-I=";
+		enum stringImportMark = "-J=";
+		enum fileImportMark = "-i=";
```
Comment (WebFreak001) [inline source/workspaced/com/ccdb.d]: can the compile_commands.json also use the shortened syntax? (`-I/path/to/dir`) - Might need to adjust the code below if so.

File: source/workspaced/com/ccdb.d [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+		enum versionMark = "-version=";
+		enum debugMark = "-debug=";
+
+		foreach (arg; args)
+		{
+			if (arg.startsWith(importMark))
```
Comment (WebFreak001) [inline source/workspaced/com/ccdb.d]: you could use the `startsWith` overload taking in multiple needles and switch on that, e.g.
```d
auto mark = arg.startsWith(importMark, stringImportMark, fileImportMark, versionMark, debugMark);
// mark is 0 if none, 1=importMark, 2=stringImportMark, 3=fileImportMark, 4=versionMark, 5=debugMark
```

I think when having so many marks it's a little better to understand and avoid copy paste errors, as well as reducing startsWith calls (which might get more optimized)

**Review by WebFreak001 — CHANGES_REQUESTED** [review]
awesome! Code quality seems pretty good, just want you to take a look at these two comments here. Good work!

File: source/workspaced/com/ccdb.d (hunk 1 above)
Comment (rtbo) [reply source/workspaced/com/ccdb.d]: The `compiled_commands.json` can use whatever syntax is accepted by the compiler.
I will fix this.

Comment (rtbo) [reply source/workspaced/com/ccdb.d]: Not obvious from the DMD manual that there are shortened syntax and for which flags.
I had to look at DMD's arg parsing code to see which flag can accept shortened version. (turns out `-I` and `-J` only)

File: source/workspaced/com/ccdb.d (hunk 2 above)
Comment (rtbo) [reply source/workspaced/com/ccdb.d]: I didn't know this overload!

**Comment by rtbo** [comment]
How do I get this kicked-in in the editor?
I've setup a VSCode workspace with `workspace-d`, `serve-d` and `code-d` to have a global view but it's far from obvious!
I can work on the 3 if needed.
The user must be able to select the `compile_commands.json` file, and pass it along to workspace-d

**Comment by WebFreak001** [comment]
serve-d registers all components and then decides which component to load on a per-workspace basis. Serve-D would need to check for the compile_commands.json to exist (like it checks if dub.json/dub.sdl exists right now) and then attempt to load the component, with fallback to other providers.

I can quickly add it to serve-d (it's a little awkward right now that I always need to bump workspace-d to a new pre-release to be able to put new stuff in serve-d)

**Review by WebFreak001 — APPROVED** [review]

**Comment by rtbo** [comment]
Ok great! Don't hesitate to ping me, I'd be happy to test.

**Comment by WebFreak001** [comment]
do you have a meson project to test this with?

**Comment by rtbo** [comment]
I've set a minimal project here:
https://github.com/rtbo/meson-d-sample

**Pull request merged**